polkadot/runtime/parachains/src/coretime/mod.rs

```diff
@@ -244,11 +288,50 @@ impl<T: Config> Pallet<T> {
 				Location::new(0, [Junction::Parachain(T::BrokerId::get())]),
 				message,
 			) {
-				log::error!("Sending `NotifyCoreCount` to coretime chain failed: {:?}", err);
+				log::error!(target: LOG_TARGET, "Sending `NotifyCoreCount` to coretime chain failed: {:?}", err);
 			}
 		}
 	}
 
+	/// Provide the amount of revenue accumulated from Instantaneous Coretime Sales from Relay-chain
+	/// block number last_until to until, not including until itself. last_until is defined as being
+	/// the until argument of the last notify_revenue message sent, or zero for the first call. If
+	/// revenue is None, this indicates that the information is no longer available. This explicitly
+	/// disregards the possibility of multiple parachains requesting and being notified of revenue
+	/// information.
+	///
+	/// The Relay-chain must be configured to ensure that only a single revenue information
+	/// destination exists.
+	pub fn notify_revenue(when: BlockNumber) -> DispatchResult {
+		let now = <frame_system::Pallet<T>>::block_number();
+		let when_bnf: BlockNumberFor<T> = when.into();
+
+		// When cannot be in the future.
+		ensure!(when_bnf <= now, Error::<T>::RequestedFutureRevenue);
+
+		let claim = <assigner_on_demand::Pallet<T>>::start_revenue_claim_until(when_bnf);
+		log::debug!(target: LOG_TARGET, "Revenue info requested: {:?}", claim.amount);
+
+		let raw_revenue: Balance = claim.amount.try_into().map_err(|_| {
+			log::error!(target: LOG_TARGET, "Converting on demand revenue for `NotifyRevenue` failed");
+			Error::<T>::AssetTransferFailed
+		})?;
+
+		with_transaction(|| -> TransactionOutcome<Result<_, DispatchError>> {
+			match do_notify_revenue::<T>(when, raw_revenue) {
+				Ok(()) => TransactionOutcome::Commit(Ok(())),
+				Err(err) => {
+					log::error!(target: LOG_TARGET, "notify_revenue failed: {err:?}");
+					TransactionOutcome::Rollback(Err(Error::<T>::AssetTransferFailed.into()))
+				},
+			}
+		})?;
+
+		<assigner_on_demand::Pallet<T>>::commit_revenue_claim(claim);
```

Comment:
Why isn't this directly committed in `start_revenue_claim_until`?

Reply:
If the XCM stuff between `start_revenue_claim_until` and `commit_revenue_claim` fails, we don't call `commit_revenue_claim`, thus leaving the revenue where it is so we can try to collect it again on the next iteration. Otherwise, we end up with funds sitting in the pot, not accounted for anywhere.

Reply:
The XCM stuff should not fail. If it fails, there is a bug.